### PR TITLE
[dynamo] "skip_guard_eval_unsafe" API for power users

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139038
* #139655

# Motivation

We have spent quite some time this year on improving guard performance and soundness. Nevertheless, guards STILL take time. We have seen multiple requests/evidences from power users where they want to have almost 0% guard overhead. First, we saw this in vLLM where even 1% overhead is bad. And recently we saw this in hqq (low precision LLM
generation) - https://github.com/pytorch/pytorch/issues/138386. To put some numbers for perspective, low precision LLM inference reaches around 250 tokens/second, i.e, each token takes a mere 4 milliseconds. If guard overhead is even 200 us, its still 5% overhead in total.

Here, users ask - "we can guarantee that there will no more recompilations in the steady state, give us the lowest guard overhead"

# Design

A must-have consideration is to support fast inference where the model has recompiled, i.e., has multiple cache entries for a code object (could be because of dynamism, or just tensor dtype change in the case of hqq). So, we still have to run the guards to figure out which compiled graph to run.

What we need is the "minimal set of differentiating guards" - i.e. minimals set of guards that we can run to choose the compiled graph. Note that this works ONLY with the assumption that users really guarantee no more recompilation scenarios (no more mutations, no more dynamism after the model has been warmed up). It is possible that if user violates this assumption, and it is not covered by the diff guard set, we will choose a wrong compiled graph to run.

When we designed C++ guards, Ed and Voz suggested to use Trie-structure to directly represent this "diff guard set". But due to complexity, we went for tree structure and relied on a GuardManager state - "fail_count" - to fail fast. I realized that we can rely on this "fail_count" to find the diff guard set.

If we recompile, this means that all the cache line guard eval check_fns have failed. Whenever a guard check_fn fails, we increment the counter in the failing node (and propagate it to the root node) to do faster fail next time. If we want to run the "guard diff set", we just have to run only those nodes in the tree which have "fail_count > 0". 

This PR relies on this observation to introduce a new stance - "skip_guard_eval". The idea is that user will warm up their model with torch.compile, and the run the steady state with this stance. This stance go through the existing cache lines for the intercepted code object but only runs the diff guard set. This dramatically reduces the guard overhead. In case, all guards fail, we fall back to eager (however if this happens then user is violating the assumption, so we should perhaps hard error, I need to fix some silly issue from _dynamo.disable to hard error here).

A bonus point here is  that this "theoretically" works with graph breaks as well. But, I need more testing to convince myself about this.

# Evaluation

I tried the hqq model in https://github.com/pytorch/pytorch/issues/138386. With very small changes in the user code ([hqq PR](https://github.com/mobiusml/hqq/pull/127)), I see the throughput increase from **160 tokens/sec to 174 tokens/sec**.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Comment by ezyang** [comment]
I'm positive on this. We should log entrance to this stance to tlparse so we can easily tell if someone is in this mode for debugging purposes. I think hard error on miss is a must. I might suggest a softer version of this that still does dynamic shapes guards (or in general, guards that are "likely" to catch correctness problems).

I'm trying to remember who was most against the sticky cache concept. Maybe @gchanan but he's not available right now. I would show it to that person to get the stiffest feedback.

**Comment by anijain2305** [comment]
> I might suggest a softer version of this that still does dynamic shapes guards (or in general, guards that are "likely" to catch correctness problems).
> 

Sounds good, this is quite easy to do today. Dynamic shape guards are sitting on the RootGuardManager, so we can always run the RootGuardManager in the "diff guard set".

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -166,6 +166,8 @@ def _callback_from_stance(callback):
     elif _stance.stance == "eager_on_recompile":
         # run mode
         return False
+    elif _stance.stance == "skip_guard_eval":
+        return torch._C._dynamo.eval_frame.skip_guard_eval_callback_flag
```
Comment (jansel) [inline torch/_dynamo/eval_frame.py:170]: I'm ok with this idea in general, but this seems like a bit of an odd way to do this, since it will also disable compiling of new frames.  Maybe this should instead be a global flag in C++ that you can control without changing stances.

We should also put unsafe in the name, and I like the idea of something softer that still checks shapes.

**Review by jansel — CHANGES_REQUESTED** [review]

**Comment by bhack** [comment]
This is interesting and I find some similarities to the `torch.export` Dims. So I think that a common overview between this two concepts it could help to expose it to the users also if we are talking about two different use case.

Can an explicit interface like `torch.export` Dims be more clear to the user for a warm-up analysis?

I think we could adopt something similar to `torch.export` logic and API when we want to skip guards but currently I found two main issue with that API?

 - Not able to pre-declare all the discrete/sparse dims I expect in my program https://github.com/pytorch/pytorch/issues/136119
 - Counterintuitive interplay between ranges and guards https://github.com/pytorch/pytorch/issues/136216

**Comment by anijain2305** [comment]
@bhack `export` is little bit out of scope here. `torch.export` is AOT while torch.compile is JIT. This completely changes how you represent and run guards.

**Comment by bhack** [comment]
> export is little bit out of scope here. torch.export is AOT while torch.compile is JIT. 

I know this

> This completely changes how you represent and run guards.

Are we sure about this also for the specific `skip_guard_eval` cases in this thread?

> Note that this works ONLY with the assumption that users really guarantee no more recompilation scenarios (no more mutations, no more dynamism after the model has been warmed up).

I know this is not exactly like the AOT case but in this specific case I see many similarities. 
E.g. we can still have cases where I need to JIT optimize the code (e.g. running on different HW instead the same hardware as with `export`) and in the warmup phase I can really expose all the possible input I expect to give to the program. 

Also talking about this warmup phase we can pass these inputs as real "live" inputs or we can always pass eventually a discrete set of a continuous range of dims right? 
As mentioned in the previous comment currently this is not supported if we consider to have an API similar to `export` Dims.

I don't know if this is still so off-topic in this specific AOT case after this a little bit expanded rationale/use cases.

I am not a lot deep inside these internals but I still see a logical connection from the public API surface and end user model develop/deployment.

**Comment by bhack** [comment]
> Note that this works ONLY with the assumption that users really guarantee no more recompilation scenarios (no more mutations, no more dynamism after the model has been warmed up).

P.s. There are also cases where this could be guaranteed, with the owner responsibility. by what were already historically collected in the remote cache for the same program and all the historically collected dynamic inputs.

**Comment by youkaichao** [comment]
just find this and chime in here. vLLM finally goes to the solution of custom dispatcher, we have some variables not seen by `torch.compile`, and we dispatch to compiled cache entries based on those variables.

see https://github.com/vllm-project/vllm/blob/main/vllm/compilation/wrapper.py and https://github.com/vllm-project/vllm/blob/91c9ebbb1bfc39e98aa2bd444b9569e5f2f92c9e/vllm/worker/tpu_model_runner.py#L680-L696

 not sure if this helps.

File: torch/compiler/__init__.py [hunk 2]
```diff
@@ -381,3 +382,30 @@ def is_dynamo_compiling() -> bool:
         >>>     # ...rest of the function...
     """
     return False
+
+
+@contextlib.contextmanager
+def skip_guard_eval_unsafe() -> Generator[None, None, None]:
```
Comment (anijain2305) [inline torch/compiler/__init__.py:388]: Bikeshed please.

**Comment by anijain2305** [comment]
@jansel This is ready for real first round of review.

File: torch/compiler/__init__.py [hunk 3]
```diff
@@ -381,3 +383,30 @@ def is_dynamo_compiling() -> bool:
...
```
Comment (ezyang) [inline torch/compiler/__init__.py:389]: @jansel are you sure about not making it a stance? I really don't see why it shouldn't be a stance, you need to hit all the compilation first before you can actually go unsafe, this seems 100% what the stance API wants to capture.

File: torch/_dynamo/guards.py [hunk 4]
```diff
@@ -292,6 +292,38 @@ def visit(mgr):
 
         visit(self.root)
 
+    def propagate_diff_guard_set(self):
```
Comment (ezyang) [inline torch/_dynamo/guards.py:295]: I am having trouble squaring the function naming with the PR description (I haven't understood the implementation yet). Your PR description suggests that you look at the "fail guard count" to find out the critical guards that are needed to distinguish various cases. But this has nothing to do with "diffing" the guards. For example, suppose I have:

"test_a() and test_b()"
"test_a() and test_c() and test_d()"

The "diff" of the guards is "test_b()" and "test_c() and test_d()". However, assuming these are the only two compiled frames, the "critical" guard set is literally "test_b()" only: there is no point in testing test_c/test_d at all, it's unnecessary (under the assumption that these are the only two compiled code you ever will actually hit.)

I'm not sure what level of safety I want out of this API. But if I want the *fastest* set of tests, there won't be any diffing involved...

File: torch/csrc/dynamo/guards.h [hunk 5]
```diff
@@ -11,6 +11,8 @@ PyObject* torch_c_dynamo_guards_init();
 // not visible there.
 void* convert_to_root_guard_manager(py::object root);
 bool run_root_guard_manager(void* root, PyObject* f_locals);
+void set_run_diff_guard_set(void* root);
+void unset_run_diff_guard_set(void* root);
```
Comment (ezyang) [inline torch/csrc/dynamo/guards.h:15]: docs would help

File: torch/csrc/dynamo/guards.cpp [hunk 6]
```diff
@@ -1879,11 +1884,28 @@ class GuardManager {
     GuardManager::add_leaf_guard(std::move(leaf_guard));
   }
 
+  void include_in_diff_guard_set() {
```
Comment (ezyang) [inline torch/csrc/dynamo/guards.cpp:1887]: docs for this and other funcs

File: torch/compiler/__init__.py [hunk 7]
```diff
@@ -381,3 +383,30 @@ def is_dynamo_compiling() -> bool:
...
+    some overhead, which may impact performance in scenarios where you need to
+    optimize for minimal guard processing time.
+
+    This API enables you to disable guard evaluation, assuming that you have
+    warmed up the compiled model with a sufficient variety of inputs. This
+    assumption means that, after the warmup phase, no further recompilations
+    will be necessary.
+
+    If this assumption fails, there is a risk of silently producing incorrect
+    results (hence the term "unsafe" in the API name).
```
Comment (ezyang) [inline torch/compiler/__init__.py:405]: There is some sort of nontrivial proof of correctness here that relies on the "trie-ness" of guard sets.

In particular, consider this near-miss counterexample. I have two compiled code with these guards:

```
a1 and a2
b1
```

Let us suppose I triggered compilation of these frames with:

```
a1=True, a2=True, b1=False
a1=False, a2=False, b1=True
```

These inputs will trigger compilation of both frames. From these two compilations, a1 will have a nonzero fail count.

When we skip guard evaluation, our adjusted guards are now:

```
a1
True
```

But this is now unsound EVEN if the user only needs to run the compiled code above. Specifically, if I have `a1=True, a2=False, b1=True`, the second graph would hit (since b1 == True), but because a1 is True I will use the first compiled code anyway.

Why is this a near-miss counterexample? Assume guards are never elided, it is impossible for the second frame to have only b1 as its guard. Instead, our guards should be:

```
a1 and a2
not a1 and b1
```

In other words, the invariant is that every compiled code must have the EXACTLY same guards, up until a "critical" guard which must be included in both guard sets but then must vary before you're allowed to now have different guards. Under these guards, `a1=True, a2=False, b1=True` isn't covered by either compiled code and so under the contract here it is OK to drop.

Unfortunately, it's difficult for me to tell if we 100% can rely on this invariant. One easy way to get into trouble is if we sometimes elide guards (which we do!) I also have some vague concern with what happens when we reorder guards but maybe it's OK because you always have to accumulate a failure on the critical guard before anything gets reordered, so the critical guards won't get dropped.

File: torch/compiler/__init__.py (hunk 3 above)
Comment (jansel) [reply torch/compiler/__init__.py:389]: We could but, I think it is orthogonal to the existing stances.  You could have:
- skip_guard_eval_unsafe + error_on_recompile
- skip_guard_eval_unsafe + eager_on_recompile
- skip_guard_eval_unsafe + default

So this would need to be a different arg.

File: torch/csrc/dynamo/extra_state.cpp [hunk 8]
```diff
@@ -142,17 +166,31 @@ void lookup(
...
       found = &cache_entry;
       break;
     }
     ++index;
   }
   if (found) {
-    extra_state->move_to_front(found);
+    if (run_diff_guards) {
+      // Undo the move_front_to_back() done at the beginning of this function.
+      extra_state->move_front_to_back();
```
Comment (ezyang) [inline torch/csrc/dynamo/extra_state.cpp:190]: How come?

File: torch/compiler/__init__.py (hunk 3 above)
Comment (ezyang) [reply torch/compiler/__init__.py:389]: Yup, sounds good to me! Separate arg.

File: torch/csrc/dynamo/eval_frame.c [hunk 9]
```diff
@@ -656,14 +675,15 @@ static PyObject* _custom_eval_frame(
...
+
   DEBUG_CHECK(PyDict_CheckExact(locals));
   DEBUG_CHECK(PyDict_CheckExact(frame->f_globals));
   DEBUG_CHECK(PyDict_CheckExact(frame->f_builtins));
 
   _PytorchRecordFunctionState* rf = _pytorch_record_function_enter(cache_lookup_profiler_str);
   PyObject* maybe_cached_code = NULL;
   const char* trace_annotation = "";
-  lookup(extra, locals, backend, &maybe_cached_code, &trace_annotation);
+  lookup(extra, locals, backend, &maybe_cached_code, &trace_annotation, PyObject_IsTrue(run_diff_guard_set));
```
Comment (ezyang) [inline torch/csrc/dynamo/eval_frame.c:686]: Why don't you just store a C++ bool as the state here lol

File: torch/csrc/dynamo/eval_frame.c [hunk 10]
```diff
@@ -524,6 +524,25 @@ static PyObject* _custom_eval_frame_shim(
 static PyObject* skip_code_recursive_flag;
 static PyObject* cache_limit_hit_flag;
 
+
+// When this flag is turned on, cache lookup runs diff guard set to find the
+// compiled graph.
+static PyObject* run_diff_guard_set;
```
Comment (jansel) [inline torch/csrc/dynamo/eval_frame.c:530]: Should this be just a bool?

File: torch/csrc/dynamo/eval_frame.c [hunk 11]
```diff
@@ -524,6 +524,25 @@ static PyObject* _custom_eval_frame_shim(
...
+
+static PyObject* set_run_diff_guard_set_flag_py(PyObject* dummy, PyObject* value) {
+  if (value != Py_True && value != Py_False) {
+    DEBUG_TRACE0("arg error");
+    PyErr_SetString(PyExc_TypeError, "expected True/False");
+    return NULL;
+  }
+  run_diff_guard_set = value;
```
Comment (jansel) [inline torch/csrc/dynamo/eval_frame.c:538]: For True/False this doesn't matter, but this is technically missing refcounts.

File: torch/csrc/dynamo/eval_frame.c [hunk 12]
```diff
@@ -524,6 +524,25 @@ static PyObject* _custom_eval_frame_shim(
...
+  Py_RETURN_NONE;
+}
+
+static PyObject* get_run_diff_guard_set_flag_py(PyObject* dummy, PyObject* args) {
+  return run_diff_guard_set;
```
Comment (jansel) [inline torch/csrc/dynamo/eval_frame.c:543]: Missing incref.  This one is actually dangerous because if you run this code enough times the refcount of True/False will reach zero.

File: torch/_dynamo/guards.py [hunk 13]
```diff
@@ -292,6 +292,38 @@ def visit(mgr):
...
+                    if val_mgr:
+                        in_diff_guard_set |= visit(val_mgr)
+            else:
+                for child_mgr in mgr.get_child_managers():
+                    in_diff_guard_set |= visit(child_mgr)
+
+            if in_diff_guard_set:
+                mgr.include_in_diff_guard_set()
+
+            return in_diff_guard_set
```
Comment (ezyang) [inline torch/_dynamo/guards.py:320]: My previous comments have been about how to handle guards treated as a list of boolean expressions evaluated in order. But we actually have trees. And now I am a little uncertain what exactly what it means to be a "critical guard" in a tree setting.... like, I don't actually want to do the parent guard tests right? If for all of my compiled code I can assume `foo[0].bar` exists, I just shouldn't bother testing if those things exist when I traverse down to the tensor right?

File: torch/csrc/dynamo/extra_state.cpp [hunk 14]
```diff
@@ -114,18 +125,31 @@ void lookup(
...
   size_t index = 0;
   CacheEntry* found = nullptr;
   py::handle locals(f_locals);
+  if (run_diff_guards && extra_state->cache_entry_list.size() > 1) {
+    // Move the first entry to the back. This is NOT a perf optimization, this
+    // is a necessity. It is possible that in the case of multiple cache
+    // entries, the first cache entry (which would be the latest recompile) does
+    // not have any guard managers with fail_count > 0. So, we just move it to
+    // the end.
+    extra_state->move_front_to_back();
```
Comment (ezyang) [inline torch/csrc/dynamo/extra_state.cpp:139]: @jansel can you remind me why we attempt the most recently compiled cache entry as opposed to the oldest cache entry lol. I guess I can imagine some reasons. Reordering guards seems very dangerous in the unsafe eval world......

File: torch/_dynamo/guards.py (hunk 4 above)
Comment (anijain2305) [reply torch/_dynamo/guards.py:295]: This is for the suggestion on the "softer" version - i.e. run atleast TENSOR_MATCH guards. So, I am marking all managers that have TENSOR_MATCH guards as include-in-the-diff-guard-set forcibly. And then I do a tree traversal to mark all the parent nodes in diff-guard-set.

**Review by jansel — CHANGES_REQUESTED** [review]
Thoughts on implementing this as a separate guard manager instance (with fewer guards) versus flags that need to be checked many times?

A separate manager might be faster, since you wouldn't need to check flags multiple times.

File: torch/csrc/dynamo/extra_state.cpp (hunk 14 above)
Comment (anijain2305) [reply torch/csrc/dynamo/extra_state.cpp:139]: MRU is useful for automatic dynamic

Comment (jansel) [reply torch/csrc/dynamo/extra_state.cpp:139]: The reordering logic is to "fail fast" where we try to check things that fail often first.  This is to mitigate the fact that we check multiple cache entries one-by-one.

I don't exactly understand why we are reordering this here though...

File: torch/compiler/__init__.py (hunk 7 above)
Comment (anijain2305) [reply torch/compiler/__init__.py:405]: Yeah, this does not work 😿 

I have to think more.

File: test/dynamo/test_skip_guard_eval_unsafe.py [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+
+        with self.assertRaisesRegex(
+            RuntimeError, "Could not find a compiled graph for skip_guard_eval_unsafe"
+        ):
+            with torch.compiler.skip_guard_eval_unsafe():
+                x = torch.randn(4, 4)
+                dt["x"] = x
+                opt_fn(dt)
+
+    def test_cache_line_pickup(self):
```
Comment (anijain2305) [inline test/dynamo/test_skip_guard_eval_unsafe.py:117]: As @ezyang pointed out [here](https://github.com/pytorch/pytorch/pull/139038#discussion_r1829849298), the reliance on fail_count is not enough. This test fails today.

For this to really work properly, we will have to add a new mode in GuardManagers that do not early return, updating fail_count for every possible failing node.

**Comment by anijain2305** [comment]
Moving to https://github.com/pytorch/pytorch/pull/140251

The main reason for closing this is implementation. In the other stack, I am introducing the facility to clone the guard manager which makes it more readable and performant.
